feat: add StringAsBytes() method and reduce garbage when SkipValue encounters string

cwaldren-ld opened the pull request:
This PR (from contributor @bobby-stripe) adds a new `StringAsBytes()` method, which can be used instead of the always-allocating `String()` method. 

This shouldn't affect LaunchDarkly's current use as the method didn't exist before, but we may be able to take advantage of this new method if we need to optimize. 

The justification for this use-case can be found [here](https://github.com/launchdarkly/go-jsonstream/pull/19).

bobby-stripe committed:
jreader: when SkipValue is called, don't allocate strings that immediately become garbage

bobby-stripe committed:
jreader: add test ensuring that SkipValue doesn't allocate strings

bobby-stripe committed:
jreader: add StringAsBytes() (#19)

When reading a new version of a large but shallow JSON file from disk
(e.g. an array of feature flag definitions), we end up creating a lot of
strings for object properties (e.g. FF name), just to end up eventually
comparing those strings to the same properties on an already-read-in
older version of a flag, and discarding the newly created ones. Exposing
`StringAsBytes()` as an alternative API to `String()` enables us to
compare new versions of flags (at Stripe we use a pre-existing, internal
fork of a feature flag library called
[`goforit`](https://github.com/stripe/goforit) to old ones without
creating garbage, as due to quirks of our distribution mechanism the
majority of flags rarely change.

Using `go-jsonstream` with this change (together with
https://github.com/launchdarkly/go-jsonstream/pull/18 , but this PR is
the most impactful) results in a 60% reduction in runtime in my usecase
(updating our feature flags from disk) compared to using
`encoding/json`, an 80% reduction in MB allocated, and a 99%(!)
reduction in objects allocated.

cwaldren-ld commented:
Hey @bobby-stripe , it looks like we need some updates to our easyjson implementation to make this work (checking the build failure in CI). I'll take a look at that, and ping you on a PR if necessary.

bobby-stripe commented:
@cwaldren-ld ah, sorry! I had forgot to test against easyjson.  Their lexer API doesn't seem to support getting a string value as bytes -- the `.Bytes()` method tries to base64 decode a string.  The best idea I can think of (which doesn't make it a _good_ idea...) is to move `Reader.String` and `Reader.StringAsBytes` implementations into `reader_default.go` and in `reader_easyjson.go` only expose `Reader.String` (using the original version that calls `tokenReader.String`), as the point of `StringAsBytes` is avoiding allocs, and I don't think we can do that with easyjson.

cwaldren-ld committed:
Merge branch 'bobby/remove-garbage' into contrib

bobby-stripe committed:
jreader: fix StringAsBytes (#22)

In the original PR with `StringAsBytes` I missed testing with easyjson.
Looking in to the implementation, there isn't a good way to get a
string-as-bytes without an allocation from `easyjson` (the easyjson
lexer _does_ have a `Bytes()` method, but its designed for decoding a
string as base64 bytes:
[link](https://github.com/mailru/easyjson/blob/v0.7.6/jlexer/lexer.go#L697)).
So while this API doesn't _help_ in the easyjson case, it isn't worse
than a caller who needs bytes doing the cast on their own, and allows
easy testing without code changes if just the build tags are flipped.

cc @cwaldren-ld

bobby-stripe committed:
jreader: address PR suggestions and fix lint build (#23)

Adds comments to `isEasyJSON` constant.

bobby-stripe commented:
💀 
```go
jreader/token_reader_default.go:22:7: const `isEasyJSON` is unused (unused)
const isEasyJSON = false //nolint:deadcode
```

@cwaldren-ld ok I'll open up a PR I guess with that other idea, moving these to test-only files.  I also had ran golangci-lint 1.50 locally rahter than what your CI is running (1.48), so I'll take a few minutes first to try to reproduce that failure here

bobby-stripe committed:
jreader: move isEasyJSON constant into test-only file to satisfy linter (#24)

bobby-stripe commented:
in addition to the `jreader.Reader.StringAsBytes()` API addition, this branch also has the ["eliminate garbage when `SkipValue` encounters strings" PR](https://github.com/launchdarkly/go-jsonstream/pull/18) as well, which significantly reduces garbage creation in the non-easyjson build.  Happy to answer any questions or help address feedback!

Thanks for helping get this up to the reviewable line @cwaldren-ld !

bobby-stripe commented:
@keelerm84 @cwaldren-ld any chance to get this reviewed/looked at soon?

bobby-stripe commented:
\o/ thanks so much @cwaldren-ld and @keelerm84 !
